osu.Server.Queues.ScoreStatisticsProcessor/Processors/UserTotalPerformanceProcessor.cs

```diff
@@ -128,5 +125,38 @@ private static async Task updateGlobalRank(UserStats userStats, MySqlConnection
                     rank = userStats.rank_score_index
                 }, transaction);
         }
+
+        // [ruleset_id, [rank_score, count_users_above]]
+        private static readonly ConcurrentDictionary<int, ConcurrentDictionary<int, int>> rank_score_index_partition_cache =
+            new ConcurrentDictionary<int, ConcurrentDictionary<int, int>>();
+
+        private static async Task<int> getUserRankScoreIndex(UserStats userStats, MySqlConnection connection, MySqlTransaction? transaction, LegacyDatabaseHelper.RulesetDatabaseInfo dbInfo)
+        {
+            var rulesetCache = rank_score_index_partition_cache.GetOrAdd(dbInfo.RulesetId, _ => new ConcurrentDictionary<int, int>());
+
+            const int partition_size = 100;
+
+            int partitionCutoff = ((int)(userStats.rank_score / partition_size) + 1) * partition_size;
+
+            // Simple cache invalidation: 0.1% chance of re-querying the partition's cache. So basically every 1000 queries.
+            if (RNG.NextSingle() > 0.999) rulesetCache.Remove(partitionCutoff, out int _);
+
+            // This query is indexed as much as it can be, but for COUNT(*) across millions of rows it can still be slow.
+            // Cache the majority lookup by partitioning over `pp` reduces the per-query lookup overhead.
+            int partitionCount = rulesetCache.GetOrAdd(partitionCutoff, cutoff => connection.QuerySingle<int>(
+                $"SELECT COUNT(*) FROM {dbInfo.UserStatsTable} WHERE rank_score > @cutoff",
+                new { cutoff }, transaction: transaction));
+
+            int remainingCount = await connection.QuerySingleAsync<int>(
+                $"SELECT COUNT(*) FROM {dbInfo.UserStatsTable} WHERE rank_score BETWEEN @rankScoreCutoff AND @partitionCutoff AND user_id != @userId",
```

Comment:
Bit of a nitpick but `BETWEEN` is both-ends inclusive. Previous query was `rank_score > @rankScoreCutoff` and now it's a union of `rank_score > @partitionCutoff` and `rank_score >= @rankScoreCutoff AND rank_score <= @partitionCutoff`, netting `rank_score >= @rankScoreCutoff` in general, so it's a slight behavioural change. In practice likely won't matter much because floats but it might matter for something like zero? dunno.

If we wanted to be super proper you'd want to have

```suggestion
                $"SELECT COUNT(*) FROM {dbInfo.UserStatsTable} WHERE rank_score > @rankScoreCutoff AND rank_score <= @partitionCutoff AND user_id != @userId",
```

I believe.